### PR TITLE
enable environments in arapp.json

### LINKED ISSUE #119: arapp overrides per network
At the moment, the `arapp` file is a unique representation of the app state which becomes a problem when someone is publishing an app to many networks. The main two problems are:

- `appId`: Which needs to be in a registry available in the network that the dev has permissions to publish to. It may vary depending on network and may change (going from `open.aragonpm.eth` to the main registry)

- `version`: Which needs to be cohesive with what versions have been previously published, else an 'invalid bump' error can occur. For the `aragon run` command, we currently completely ignore this value and always publish a new major version as a workaround. On this issue, what is the reason to have the version number in package description files such as `package.json` (cc @sohkai @bpierre) besides being able to process them in a server? I'm thinking that we could drop the version value from `arapp.json` altogether, replacing it with an argument to `apm publish` in which you specify the type of bump or version number you want to publish.

My initial idea is that in the truffle configuration file, we could read an `arapp` key per network and then override those values.

### PR DESCRIPTION
**Pull request by perissology**
resolves #119

I think keeping the `aragon apm version` cmd would be nice to show the current published version. However I decided to deprecate it b/c having different behaviors for the same cmd depending on if the app is using the new or old `arapp` format would be confusing for users. You can also obtain the same information via `aragon apm versions`

Note: that some `yargs` provided behavior (ex. `defaults`, `exclusion`) is handled in the `environmentMiddleware` b/c the `yargs` functionality doesn't really work very well when deprecating a command

### CODE REVIEW

File: src/middleware/environment.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    const env = module.environments[environment]
+
+    if (!env) {
+      reporter.error(`${environment} environment was not defined in your arapp.json.`)
+      process.exit(1)
+    }
+
+    const resp = {
+      module: Object.assign({}, module, { appName: env.appName }),
```
Comment (izqui) [inline src/middleware/environment.js]: Maybe we should extract all the other environments from the module before returning it, so it isn't part of the `artifact.json` when it gets generated on publish.

**Comment by perissology** [comment]
> It would be great to do a PR to one of the boilerplates (https://github.com/aragon/aragon-react-kit-boilerplate) with the new arapp config so we can test to init projects with the new config.

`aragon init` will convert the old `arapp.json` in the template to the new format automatically, so no need to update the boilerplate before using

File: src/commands/apm_cmds/publish.js [hunk 2]
```diff
@@ -24,13 +24,16 @@ const getRepoTask = require('../dao_cmds/utils/getRepoTask')
...
 
-exports.command = 'publish [contract]'
+exports.command = 'publish <bump> [contract]'
 
 exports.describe = 'Publish a new version of the application'
 
 exports.builder = function (yargs) {
   return deploy.builder(yargs) // inherit deploy options
-    .positional('contract', {
+    .positional('bump', {
```
[the remark this answers is not in the document]
Comment (perissology) [reply src/commands/apm_cmds/publish.js]: Ya, I can merge that in. I do think it makes more sense to make `bump` a required positional arg, then a flag. Thoughts?

File: src/middleware/environment.js (hunk 1 above)
Comment (perissology) [reply src/middleware/environment.js]: implemented in L87

File: src/middleware/environment.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      network: configureNetwork(argv, env.network)
+    }
+    if (env.registry) {
+      resp.apmEnsRegistry = env.registry
+      if (apm) {
+        apm['ens-registry'] = env.registry
```
[the remark this answers is not in the document]
Comment (perissology) [reply src/middleware/environment.js]: I updated `cli.js` and made a [PR](https://github.com/aragon/hack/pull/45) for the docs.

I'll make a separate PR to clean up the `ensRegistryAddress`

File: src/middleware/environment.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+const configureNetwork = (argv, network) => {
+  // Catch commands that dont require network and return
+  const skipNetworkSubcommands = new Set(['version']) // 'aragon apm version'
+  if (argv.length >= 4) {
+    if (skipNetworkSubcommands.has(argv[3])) {
+      return {}
+    }
+  }
+
+  const skipNetworkCommands = new Set(['init', 'devchain', 'ipfs'])
```
Comment (izqui) [inline src/middleware/environment.js]: Hey

File: src/commands/apm_cmds/publish.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (izqui) [reply src/commands/apm_cmds/publish.js]: Makes sense

**Comment by izqui** [comment]
🙏🚀 Thanks for getting this done, @perissology. Such an amazing improvement to the developer experience!

**Pull request merged**